### PR TITLE
[Clang][Interp] Diagnose use sizeless vector type as the argument of `__builtin_vectorelements`

### PR DESCRIPTION
**Pull request by yronglin**
This PR implement an opcode to diagnose use sizeless vector type as the argument of `__builtin_vectorelements`.

### CODE REVIEW

File: clang/lib/AST/Interp/Compiler.cpp [hunk 1]
```diff
@@ -1698,10 +1698,8 @@ bool Compiler<Emitter>::VisitUnaryExprOrTypeTraitExpr(
   if (Kind == UETT_VectorElements) {
     if (const auto *VT = E->getTypeOfArgument()->getAs<VectorType>())
       return this->emitConst(VT->getNumElements(), E);
-
-    // FIXME: Apparently we need to catch the fact that a sizeless vector type
-    // has been passed and diagnose that (at run time).
-    assert(E->getTypeOfArgument()->isSizelessVectorType());
+    if (E->getTypeOfArgument()->isSizelessVectorType())
+      return this->emitSizelessVectorElementSize(E);
```
Comment (tbaederr) [inline clang/lib/AST/Interp/Compiler.cpp:1702]:
```suggestion
    assert(E->getTypeOfArgument()->isSizelessVectorType());
    return this->emitSizelessVectorElementSize(E);
```

File: clang/lib/AST/Interp/Interp.h [hunk 2]
```diff
@@ -29,6 +29,7 @@
 #include "State.h"
 #include "clang/AST/ASTContext.h"
 #include "clang/AST/Expr.h"
+#include "clang/Basic/DiagnosticSema.h"
```
Comment (tbaederr) [inline clang/lib/AST/Interp/Interp.h:32]: What do we need this include for?

File: clang/lib/AST/Interp/Interp.h [hunk 3]
```diff
@@ -2735,6 +2736,13 @@ inline bool InvalidDeclRef(InterpState &S, CodePtr OpPC,
   return CheckDeclRef(S, OpPC, DR);
 }
 
+inline bool SizelessVectorElementSize(InterpState &S, CodePtr OpPC) {
```
Comment (tbaederr) [inline clang/lib/AST/Interp/Interp.h:2739]: If you look at what the current interpreter does: 1) This should use `CCEDiag` instead and 2) the diagnostic should only be emitted if we're in a constant context

File: clang/lib/AST/Interp/Compiler.cpp (hunk 1 above)
Comment (yronglin) [reply clang/lib/AST/Interp/Compiler.cpp]: Done

File: clang/lib/AST/Interp/Interp.h (hunk 2 above)
Comment (yronglin) [reply clang/lib/AST/Interp/Interp.h]: Removed.

File: clang/lib/AST/Interp/Interp.h (hunk 3 above)
Comment (yronglin) [reply clang/lib/AST/Interp/Interp.h:2738]: Thanks, agree 100%, it should keeps same behavior with https://github.com/llvm/llvm-project/blob/c8c0b18b5d58415b79ea13b62eee70c130c26a0a/clang/lib/AST/ExprConstant.cpp#L14080-L14092 .

**Review by tbaederr — APPROVED** [review]
LGTM, thanks!

**Pull request merged**